### PR TITLE
Migration queries refactorization - 084 to 080

### PR DESCRIPTION
**Pull request by AdrienClairembault**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes

### CODE REVIEW

**Comment by AdrienClairembault** [comment]
The tests return the following warning :
`DISTINCT selection can only take one field`
Why is multiple fields selection not allowed ?

**Comment by AdrienClairembault** [comment]
This should not be merge before https://github.com/glpi-project/glpi/pull/5190 so i can rebase and use the new syntax.

File: install/update_0803_083.php [hunk 1]
```diff
@@ -283,41 +288,53 @@ function update0803to083() {
                      ON (`glpi_notificationtemplates`.`id`
                            = `glpi_notificationtemplatetranslations`.`notificationtemplates_id`)
                 WHERE `glpi_notificationtemplates`.`itemtype` = 'Ticket'";
```
Comment (trasher) [inline install/update_0803_083.php]: Maybe you missed to drop the raw query

File: install/update_0803_083.php [hunk 2]
```diff
@@ -715,35 +787,42 @@ function update0803to083() {
...
+            if (count($DB->request($query)) === 0) {
+               $DB->insertOrDie("glpi_groups_users", [
+                     'users_id'     => $data['users_id'],
+                     'groups_id'    => $data['id'],
+                     'is_manager'   => 1
+                  ],
+                  "0.83 insert manager of groups"
+               );
+            } else {
+               // Update user as manager if presnet in groups_users
```
Comment (trasher) [inline install/update_0803_083.php]: Please fix typo on "present"

File: install/update_0803_083.php [hunk 3]
```diff
@@ -784,31 +864,40 @@ function update0803to083() {
...
+         ),
+         'FROM'   => "glpi_rules",
+         'WHERE' => [
+            'sub_type' => $ruletype
+         ],
+         'GROUPBY' => 'sub_type'
+      ];
+      if ($data = $DB->request($query)->next()) {
+         // Get rule string
+         $rules = explode(",", $data['ids']);
```
Comment (trasher) [inline install/update_0803_083.php]: Use directly `$data['ids']`

File: install/update_0803_083.php [hunk 4]
```diff
@@ -990,19 +1090,34 @@ function update0803to083() {
...
-                          FROM `glpi_tickets`)";
-   $DB->queryOrDie($query, "0.83 clean glpi_slalevels_tickets");
+   $DB->deleteOrDie("glpi_slalevels_tickets", [
+         'NOT' => [
+            'glpi_slalevels_tickets.tickets_id' => new \QuerySubQuery([
+               'SELECT' => "glpi_tickets.id",
+               'FROM'   => "glpi_tickets"
+            ])
+         ],
+         'NOT' => [
```
Comment (trasher) [inline install/update_0803_083.php]: This won't work. Yoou have two keys with the same name, php will retrieve only one.

File: install/update_0803_083.php [hunk 5]
```diff
@@ -1307,26 +1430,37 @@ function update0803to083() {
...
+         // Grab helpdesk profiles
+         $helpdesk_profiles = [];
+         $query = [
+            'FROM'   => "glpi_profiles",
+            'WHERE'  => [
+               'interface'       => "helpdesk",
+               'reminder_public' => "r"
+            ]
+         ];
+         foreach ($DB->request($query) as $data2) {
```
Comment (trasher) [inline install/update_0803_083.php]: Dot not use DB::request in loops

File: install/update_0831_084.php [hunk 6]
```diff
@@ -1201,57 +1361,84 @@ function update0831to084() {
...
-            foreach ($tab as $old => $new) {
-               $query = "UPDATE `glpi_rulecriterias`
-                         SET `criteria` = '$new'
-                         WHERE `criteria` = '$old'
-                               AND `rules_id` IN ($rules)";
-               $DB->queryOrDie($query, "0.84 update datas for rules criteria");
-            }
+      $data = $DB->request([
+         'SELECT' => new \QueryExpression(
+            "GROUP_CONCAT(" . DBmysql::quoteName('id') .") AS rules_id"
```
Comment (cedric-anne) [inline install/update_0831_084.php]:
```suggestion
            "GROUP_CONCAT(" . DBmysql::quoteName('id') .") AS (" . DBmysql::quoteName('rules_id')
```

File: install/update_0831_084.php [hunk 7]
```diff
@@ -1201,57 +1361,84 @@ function update0831to084() {
...
+         "glpi_ruleactions.id AS id",
+         "sub_type",
+         "glpi_manufacturers.name AS newvalue"
+      ],
+      'FROM'         => "glpi_rules",
+      'INNER JOIN'   => [
+         'glpi_ruleactions' => [
+            'ON' => [
+               'glpi_rules'      => "id",
+               'glpi_ruleactions' => "rules_id"
```
Comment (cedric-anne) [inline install/update_0831_084.php]:
```suggestion
               'glpi_ruleactions' => "rules_id",
               [
                  'AND' => ['field' => 'Manufacturer']
               ]
```

File: install/update_0831_084.php [hunk 8]
```diff
@@ -1201,57 +1361,84 @@ function update0831to084() {
...
+      'LEFT JOIN'    => [
+         'glpi_manufacturers' => [
+            'ON' => [
+               'glpi_manufacturers' => "id",
+               'glpi_ruleactions'   => "value"
+            ]
+         ]
+      ],
+      'WHERE'        => [
+         'field'     => "Manufacturer",
```
Comment (cedric-anne) [inline install/update_0831_084.php]:
```suggestion
```

File: install/update_0831_084.php [hunk 9]
```diff
@@ -2141,18 +2457,23 @@ function updateNetworkFramework(&$ADDTODISPLAYPREF) {
...
                                       'Index mac field and transform address mac to lower'));
 
-   $query = "UPDATE `glpi_networkports`
-             SET `mac` = LOWER(`mac`)";
-   $DB->queryOrDie($query, "0.84 transforme MAC to lower case");
+   $DB->updateOrDie("glpi_networkports", [
+         'mac' => new \QueryExpression("LOWER(" . DBmysql::quoteName("mac") .")")
+      ],
+      [true],
+      "0.84 transforme MAC to lower case"
```
Comment (cedric-anne) [inline install/update_0831_084.php]:
```suggestion
      "0.84 transform MAC to lower case"
```

File: install/update_0831_084.php [hunk 10]
```diff
@@ -2262,10 +2583,13 @@ function updateNetworkFramework(&$ADDTODISPLAYPREF) {
       $DB->queryOrDie($query, "0.84 create glpi_networkportaggregates");
 
       // Transform NetworkEquipment local MAC address as a networkport that aggregates all ports
-      $query = "SELECT *
-                FROM `origin_glpi_networkequipments`
-                WHERE `mac` != ''
-                      OR `ip` != ''";
+      $query = [
+         'FROM'   => "origin_glpi_networkequipments",
+         'WHERE'  => [
```
Comment (cedric-anne) [inline install/update_0831_084.php]: `OR` is missing.

File: install/update_0831_084.php [hunk 11]
```diff
@@ -2353,31 +2685,37 @@ function updateNetworkFramework(&$ADDTODISPLAYPREF) {
...
-               $DB->query($query);
-               unset($query);
-            }
+         $ipaddress_result = $DB->request([
+            'SELECT' => "id",
+            'FROM'   => 'glpi_ipaddresses',
+            'WHERE'  => [
+               new \QueryExpression(
+                  "(". DBmysql::quoteName("glpi_ipaddresses.binary_3") . "& " .
+                  DBmysql::quoteValue($netmask) . ")"
```
Comment (cedric-anne) [inline install/update_0831_084.php]:
```suggestion
                  DBmysql::quoteValue($netmask) . ")  = " . DBmysql::quoteValue($address)
```

File: install/update_0803_083.php (hunk 5 above)
Comment (trasher) [reply install/update_0803_083.php]: (this one for the entire PR, I won't comment each one ;))

File: install/update_080_0801.php [hunk 12]
```diff
@@ -38,39 +38,44 @@
...
-               }
-            }
-         }
+   $query = [
+      'SELECT'    => [
+         "tickets_id",
+         "type",
+         "groups_id"
+      ],
+      'COUNT'     => "CPT",
```
Comment (trasher) [inline install/update_080_0801.php]: I'm not sure this one will give expected results, this should be checked to be sure

File: install/update_0831_084.php [hunk 13]
```diff
@@ -212,25 +236,30 @@ function update0831to084() {
...
+         $rules = $rulesIterator->next()['ids'];
+
+         // Update actions
+         foreach ($status as $old => $new) {
+            $DB->updateOrDie('glpi_ruleactions', [
+                  'value' => $new
+               ], [
+                  'field'     => $field,
+                  'value'     => $old,
+                  'rules_id'  => explode(",", $rules)
```
Comment (trasher) [inline install/update_0831_084.php]: Not sure we ant implode here

Comment (AdrienClairembault) [reply install/update_0831_084.php]: I think we need an array to get an "IN => (a,b,c,...)" criteria, is there a better way to do this ?

Comment (trasher) [reply install/update_0831_084.php]: Yes indeed, an array is needed. I did not pay attention you get a string with commas from the DB, so this is ok

File: install/update_0803_083.php (hunk 4 above)
Comment (AdrienClairembault) [reply install/update_0803_083.php]: I've replaced it by a QueyExpression to get the exact same behavior.

File: install/update_0803_083.php (hunk 3 above)
Comment (trasher) [reply install/update_0803_083.php]: Indeed

File: install/update_0831_084.php (hunk 11 above)
Comment (cedric-anne) [reply install/update_0831_084.php]: You cannot use a QueryExpression as array key, you have to put the `  = " . DBmysql::quoteValue($address)` inside QueryExpression.

File: install/update_0803_083.php [hunk 14]
```diff
@@ -1104,14 +1098,14 @@ function update0803to083() {
...
          'NOT' => [
-            'glpi_slalevels_tickets.tickets_id' => new \QuerySubQuery([
-               'SELECT' => "glpi_tickets.id",
-               'FROM'   => "glpi_tickets"
-            ])
-         ],
-         'NOT' => [
-            'glpi_slalevels_tickets.slalevels_id' => new \QuerySubQuery([
-               'SELECT' => "glpi_tickets.slalevels_id",
+            new \QueryExpression(
```
Comment (cedric-anne) [inline install/update_0803_083.php]: You cannot use a QueryExpression as array key, you have to put the `IN ...` inside the QueryExpression. If you want, you can build it using a QuerySubQuery `getQuery()` method.

File: install/update_0831_084.php [hunk 15]
```diff
@@ -1913,15 +2199,25 @@ function updateNetworkFramework(&$ADDTODISPLAYPREF) {
...
+            ),
+            "netmask",
+            "gateway",
+            "entities_id"
+         ],
+         'DISTINCT'  => true,
+         'FROM'      => "origin_glpi_networkports",
+         'ORDER'     => "gateway DESC"
+      ]);
       $address = new IPAddress();
```
Comment (cedric-anne) [inline install/update_0831_084.php]: Not mandatory, but this variable is not used (redefined later for another usage).
```suggestion
```

**Pull request merged**